### PR TITLE
Add Azure Integrator support to k8s charms

### PR DESCRIPTION
**Pull request by johnsca**
Adds Microsoft Azure native integration via the azure-integrator charm.

Also tries to defer starting components if cloud integration is pending but not yet complete, to avoid unnecessary restarts.

Depends on juju-solutions/charm-azure-integrator#15 and juju-solutions/interface-azure-integration#4

### CODE REVIEW

**Comment by johnsca** [comment]
Note, I refactored the status handling of the worker a bit to better match how the master does things, with a finalizer that handles the status in one place.  Please check that I didn't break any conditions and make the status inaccurate.

File: cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py [hunk 1]
```diff
@@ -892,10 +908,10 @@ def launch_default_ingress_controller():
         'ingress-ssl-chain-completion')
     context['ingress_image'] = config.get('nginx-image')
     if context['ingress_image'] == "" or context['ingress_image'] == "auto":
-        images = {'amd64': 'quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.16.1', # noqa
-                  'arm64': 'quay.io/kubernetes-ingress-controller/nginx-ingress-controller-arm64:0.16.1', # noqa
-                  's390x': 'quay.io/kubernetes-ingress-controller/nginx-ingress-controller-s390x:0.16.1', # noqa
-                  'ppc64el': 'quay.io/kubernetes-ingress-controller/nginx-ingress-controller-ppc64le:0.16.1', # noqa
+        images = {'amd64': 'quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.16.1',  # noqa
```
Comment (hyperbolic2346) [inline cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: why two spaces?

Comment (johnsca) [reply cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: That's what my linter (flake8) said was right according to some PEP or other.

**Comment by Cynerva** [comment]
I'm finding the addition of `kubernetes-master.cloud.pending`, `kubernetes-master.cloud.blocked`, and `kubernetes-master.cloud.ready` to make this code a lot more difficult to follow.

For example, it took me a good 10-15 minutes to dig up code in charms.reactive ([here](https://github.com/juju-solutions/charms.reactive/blob/master/charms/reactive/bus.py#L352)) and convince myself that `set_cloud_pending` can't possibly run when `kubernetes-master.cloud.ready` is set, and that's only true because the handlers setting `kubernetes-master.cloud.ready` also happen to be removing other, unrelated flags.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 2]
```diff
@@ -1639,7 +1647,25 @@ def clear_cluster_tag_sent():
...
+@when_not('kubernetes-master.cloud.ready')
+def set_cloud_pending():
+    k8s_version = get_version('kube-apiserver')
+    k8s_1_11 = k8s_version >= (1, 11)
+    k8s_1_12 = k8s_version >= (1, 12)
+    vsphere_joined = is_state('endpoint.vsphere.joined')
+    azure_joined = is_state('endpoint.azure.joined')
+    if (vsphere_joined and not k8s_1_12) or (azure_joined and not k8s_1_11):
+        set_state('kubernetes-master.cloud.blocked')
+    set_state('kubernetes-master.cloud.pending')
```
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Could this be preempted by the `cloud_ready` handler? This handler might never get a chance to run, which would prevent `kubernetes-master.cloud.blocked` from being set.

Comment (Cynerva) [reply cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Ah, no, I missed that `cloud_ready` is gated on `endpoint.*.ready`, not `endpoint.*.joined`. This is probably OK.

**Comment by Cynerva** [comment]
> Note, I refactored the status handling of the worker a bit to better match how the master does things, with a finalizer that handles the status in one place. Please check that I didn't break any conditions and make the status inaccurate.

Thanks. I've been wanting to do this for a long time. :+1:

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py (hunk 2 above)
Comment (Cynerva) [reply cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Actually, is it still a concern for interfaces that don't have a clear `joined` -> request integration -> `ready` flow? I'm thinking of openstack and vsphere in particular.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 3]
```diff
@@ -1675,18 +1709,27 @@ def request_integration():
 
 
 @when_none('endpoint.aws.joined',
-           'endpoint.gcp.joined')
+           'endpoint.gcp.joined',
+           'endpoint.openstack.joined',
+           'endpoint.vsphere.joined',
+           'endpoint.azure.joined')
 @when('kubernetes-master.cloud-request-sent')
 def clear_requested_integration():
```
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Is the `@when('kubernetes-master.cloud-request-sent')` still appropriate for this handler? It seems like we might want this to get called when departing from openstack or vsphere, where `kubernetes-master.cloud-request-sent` never gets set.

**Comment by Cynerva** [comment]
If I saw a charm status saying `Azure integration requires K8s 1.11 or greater`, one thing I might try to do is upgrade to Kubernetes 1.11. However, it looks like the charm would not clear the `kubernetes-master.cloud.blocked` flag if I did this.

So if I tried to upgrade to Kubernetes 1.11, I would expect to hit two problems:
1. The charms would not run `start_master`, which would prevent the newly upgraded snaps from being reconfigured properly.
2. The `Azure integration requires K8s 1.11 or greater` status message would still show even though I've upgraded to 1.11.

I could work around this by removing and re-adding the cloud relation, but the charm doesn't communicate that to the user very well.

Same concern applies to `kubernetes-worker.cloud.blocked` and `start_worker` also.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py (hunk 3 above)
Comment (Cynerva) [reply cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: This stands out as a problem for vsphere in particular. I don't see any way for `kubernetes-master.cloud.blocked` to get cleared after relating to vsphere.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 4]
```diff
@@ -1675,18 +1709,27 @@ def request_integration():
...
           'endpoint.gcp.ready',
           'endpoint.openstack.ready',
-          'endpoint.vsphere.ready')
-@when_not('kubernetes-master.restarted-for-cloud')
-def restart_for_cloud():
+          'endpoint.vsphere.ready',
+          'endpoint.azure.ready')
+@when_not('kubernetes-master.cloud.blocked',
+          'kubernetes-master.cloud.ready',
+          'kubernetes-master.restarted-for-cloud')  # compat. TODO: remove
```
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Might be worthwhile to handle the migration from `kubernetes-master.restarted-for-cloud` to `kubernetes-master.cloud.ready` in the `upgrade-charm` hook so you don't have to do this.

**Comment by johnsca** [comment]
There seems to be a few more fields that need to be set in the cloud config file to make LBs work properly.  Working on that.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py (hunk 3 above)
Comment (johnsca) [reply cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Yeah, I changed that in the worker and forgot to do the same for the master.

File: cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py [hunk 5]
```diff
@@ -64,6 +64,15 @@
...
 @hook('upgrade-charm')
 def upgrade_charm():
+    # migrate to new flags
+    if is_state('kubernetes-worker.restarted-for-cloud'):
+        remove_state('kubernetes-worker.restarted-for-cloud')
+        set_state('kubernetes-worker.cloud.ready')
+    if is_state('kubernetes-worker.cloud-request-sent'):
+        # minor change, just for consistency
+        remove_state('kubernetes-worker.cloud-request-sent')
+        set_state('kubernetes-worker.cloud.request-sent')
```
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: Looks like `kubernetes-worker.cloud-request-sent` is still referenced in a couple other places as well. Can you update those?

**Comment by Cynerva** [comment]
@johnsca Looks like you've cleared up most of my concerns nicely, thanks. Just had one last concern pop up about the `cloud-request-sent` flags (previous 2 comments) but otherwise this looks good :+1:

**Pull request merged**